### PR TITLE
Zahary's improvements

### PR DESCRIPTION
**Pull request by Araq**
- Many improvements to the concept feature.
- Introduce 'vtptr' and 'vtref' system types as documented in the manual.

### CODE REVIEW

File: compiler/semmagic.nim [hunk 1]
```diff
@@ -128,19 +132,27 @@ proc evalTypeTrait(traitCall: PNode, operand: PType, context: PSym): PNode =
     return typeWithSonsResult(tyNot, @[operand])
   of "name":
     result = newStrNode(nkStrLit, operand.typeToString(preferTypeName))
-    result.typ = newType(tyString, context)
+    result.typ = newType(tyString, owner)
     result.info = traitCall.info
+  of "vtptr", "vtref":
+    debug operand
```
Comment (Araq) [inline compiler/semmagic.nim]: Don't forget to remove this.

Comment (zah) [reply compiler/semmagic.nim]: This is still work-in-progress. It's not ready for merging yet.

Comment (Araq) [reply compiler/semmagic.nim]: The bugfixes are blocking the release though.

Comment (dom96) [reply compiler/semmagic.nim]: Don't pressure poor @zah :)

**Comment by Araq** [comment]
Cherry-picked the important bugfixes.

**Pull request closed without merging**